### DISCUSSION TITLE
Do not convert to Stage when flattening with tusdcat

### DISCUSSION THREAD

**Discussion by rdeioris** [syoyo/tinyusdz#204] [Ideas]
Hi, currently when doing composition tusdcat is basically broken as the construction from PrimSpec to Prim is not recursive

```cpp
  for (const auto &primspec : layer.primspecs()) {
     if (auto pv = detail::ReconstructPrimFromPrimSpec(primspec.second, warn, err)) {
       stage.add_root_prim(std::move(pv.value()));
     }
   }

```

I have already a fix for it (going to open an issue and a pull request) but is it necessary to transform the final composite layer to a stage ? The goal of tusdcat is to print the usda representation, there is no need to interpret every PrimSpec to a Prim.

Should not the to_string() call to the final composite layer more than enough? (especially because when transforming a layer to a Stage you can potentially lose properties or other informations present in the original file)

**Comment by syoyo** [comment]
Yes, it is preferable to use  Layer/PrimSpec(after composition) to print USD content in flatten mode, since (flattened) Layer to Stage conversion is a redundant operation.

> when transforming a layer to a Stage you can potentially lose properties or other informations

 No, information loss should not happen from a Layer to a Stage(and vice versa) , since Stage/Prim is the 'typed' state of Layer/PrimSpec(and Stage is immutable  to USD composition operations) . 

If there is a difference between Stage <-> Layer conversion(with NO composition ops involved), there will be an issue/bug in TinyUSDZ)
